[WIP] Add warning when sending message to announcement streams

imnitishng opened the pull request:
Normal users are not allowed to send messages in announcement streams,
currently when an unauthorised user sends message to this stream, nothing happens.
This commit changes this behaviour and adds a footer warning for the same.

Fixes: https://github.com/zulip/zulip-terminal/issues/682

imnitishng commented:
Hi, maintainers.
This is my first PR for the repository and terminal based applications with python is pretty new for me, so can I have a small review for this PR to point out the mistakes, because I believe this might not be the best way to implement this feature, also I have not written any tests for this. 
So there is a bit of work to do, I will complete it after the review. Thanks. :smile:

kaustubh-nair commented:
@imnitishng thanks for the PR!
You've got the sense of the issue correctly, and the solution does seem to work OK at a quick glance. 
The way the error is displayed can be modified though - instead of adding the error to `WriteBox` the message can be displayed in the footer using `view.set_footer_text()`
Also, `WriteBox` should not open for this, i.e the user should not have the option to type a message at all.
If you want to discuss further, feel free to chat on https://chat.zulip.org/#narrow/stream/206-zulip-terminal

neiljp commented:
@imnitishng You seem to have the right broad idea here, but as @kaustubh-nair mentions, there is a pre-existing method we likely want to use here to set the text in the footer (for a short time). The check likely wants to be triggered upon any composing keypress.

imnitishng commented:
Hi @kaustubh-nair @neiljp, thanks a lot for the help.
I believe we're good to go now? Please have a look.

neiljp reviewed (commented):
@imnitishng This looks to use the existing code more, but after a little manual testing it seems to get stuck. I've given a pointer as to why this is the case and how you might move the code around to fix it.

Note that you can place the 'fixes' line at the end of the commit too, and it will close the issue when the commit is merged :)

We don't seem to have the stream property you're using here in any test or fixtures yet, so if you wanted to work on that at the same time or afterwards you'd be welcome to do that. I've not checked which version of the server added the stream property we're using here, but we may need to think about that and check; we're currently focusing on targeting version 2.1 as the minimum server version.

imnitishng commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -60,36 +60,42 @@ def private_box_view(self, button: Any=None, email: str='') -> None:
+        if self.model.stream_dict[self.model.stream_id].get(
+                'is_announcement_only', None) == 1:
+            self.msg_footer = self.view.set_footer_text(
+                "You are not authorised to send messages to this stream.", 5)
```
>  I'm surprised you didn't notice the interface being unresponsive after this point, until eg. you press esc.

I did notice it, but I thought it would be normal to Escape out of the warning.
But thank you for the pointer, I have fixed by moving the check below.

kaustubh-nair replied:
Instead of doing this inside `stream_box_view` can't we add a check to `MessageBox`'s `keypress` method?

imnitishng replied:
Neil proposed 2 ways so I did it this way. I do see few `footer` warnings called in other class' `keypress()` methods. 
I will move the check as you said.

imnitishng commented:
> so if you wanted to work on that at the same time or afterwards you'd be welcome to do that.

Sure I would love to add some tests for this PR and other missed methods.

> I've not checked which version of the server added the stream property we're using here, but we may need to think about that and check; we're currently focusing on targeting version 2.1 as the minimum server version.

I went through the [changelogs](https://zulipchat.com/api/changelog) of Zulip API and found that the attribute `is_announcement_only` that we are using for this check is deprecated in Zulip 2.2
![Screenshot from 2020-06-21 10-34-46](https://user-images.githubusercontent.com/35267629/85217175-8297fe00-b3ab-11ea-93eb-3bb31a08f2bf.png)
I think I should change the checking logic and use `stream_post_policy` instead, also add tests using the same attribute.
Please confirm. Thank You

kaustubh-nair commented:
Ah, good job noticing that - I think using `stream_post_policy` is a better idea :+1:
I'd recommend holding off adding tests to `WriteBox`/`MessageBox` since it would conflict with some existing PRs (Like #675 )

imnitishng commented:
> maybe we'd have to use both?

I have updated the PR keeping this in mind. Please have a look.

kaustubh-nair commented:
I think this will be simpler once #653 is merged, so we can check directly based on server versions

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -751,10 +751,23 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
-                    title=self.message['subject']
-                )
+                # Check if user is allowed to post
+                if (self.model.stream_dict[self.message.get('stream_id')].
+                        get('is_announcement_only', None) == 1
+                    or (self.model.initial_data.get('is_admin', None)
+                    or self.model.initial_data.get('is_owner', None)
+                    and self.model.stream_dict[self.message.get('stream_id')].
+                        get('stream_post_policy', None) == 2)):
```
`is_announcement_only` is still present in what will soon be `3.0`, so we could stick with just it for now to begin with, since it works across the versions we plan to support.

If we support the new versions, we should likely have this logic in the model, in any case.

We'd ideally want tests against the two different server versions (or feature levels). While we added some background for this in #653, we don't have good examples of using this in a feature like this yet, so you're welcome to give it a go but you may wish to discuss on the stream about this first.

neiljp reviewed (commented):
@imnitishng This seems to work fine in my manual testing, in terms of limiting myself posting as a non-admin :+1:

I'm happy to merge a simple version similar to this, but if we look for the next version features then we should consider how to structure and test it further. Let us know how you far you'd like to take this :+1:

preetmishra reviewed (commented):
@imnitishng Thanks for working on this. I ran your changes locally. It seems to work well for the case when a message is directly being sent from the 'announce' stream. :+1: 

However, when the compose box is already open, in some other stream, and I am trying to send a message to the 'announce' stream, there is no feedback. We may want to have a footer update for this case as well.

neiljp commented:
@imnitishng Just checking with you if you plan to take this further in the near term, as this is something we'd like to integrate soon.

imnitishng commented:
Hi @neiljp, I am occupied as of now so someone else can take this up.

neiljp commented:
@imnitishng I've added WIP here to indicate the status, though you're welcome to take this up again at some point of course :)

imnitishng committed:
helper.py: Add method `unauthorised_warning` to update footer

This method acts as an sanitary check when a normal user is sending
message to a stream where only organization admins are allowed to
send. The method also updates the footer with appropriate warning.

imnitishng committed:
boxes.py: Add warning when user without permission sends message

This commit adds warning for when an unauthorised user tries to
send a message to an admin only stream.
